riscos_toolbox/__init__.py

```diff
@@ -1,190 +1,140 @@
-
-def quit():
-     global _quit
-     _quit = True
-
-def run():
-    poll_block = swi.block(64)
+               appdir, _msgtrans_block, ctypes.addressof(_id_block))
+
+def msgtrans_lookup(token, *args, bufsize=256):
+    args = args[:4]
+    buffer = swi.block(int((bufsize+3)/4))
+    swi.swi("MessageTrans_Lookup","bsbi"+("s"*len(args)),
+            _msgtrans_block, token, buffer,bufsize, *args)
+    return buffer.ctrlstring()
+
+def extract_gadget_info_from_template(self, template, gadget):
+    blk = swi.swi("Toolbox_TemplateLookUp", "0s;I", template)
+    data,size = swi.swi("Window_ExtraxtGadgetInfo","iii;II",
+                        0, self.id, gadget)
+
+def run(application):
+    poll_buffer = (ctypes.c_byte*256)()
     global _quit
 
     while not _quit:
-        reason,sender = swi.swi('Wimp_Poll','Ib;I.I', 0b1, poll_block)
-
-        if reason == 0x200: # Toolbox Event
-            size       = poll_block[0]
-            reference  = poll_block[1]
-            event_code = poll_block[2]
-            flags      = poll_block[3]
-
-            if event_code == 0x44ec1: # Object_AutoCreated
-                name      = poll_block.nullstring(0x10,size)
-                obj_class = swi.swi('Toolbox_GetObjectClass', '0I;I',
-                                    _id_block.self.id)
-                _objects[_id_block.self.id] = \
-                     Object.create(obj_class, name, _id_block.self.id)
-                continue
-
-            object  = get_object(_id_block.self.id)
-            comp_id = _id_block.self.component
-
-            if object:
-                if comp_id in object.components:
-                    component = object.components[comp_id]
-                    component.event_handler(event_code, _id_block, poll_block)
-                else:
-                    object.event_handler(event_code, _id_block, poll_block)
-
-        if reason == 2: # Open Window
-            if _id_block.self.id in _objects:
-                object = _objects[_id_block.self.id]
-                object.wimp_handler(reason, _id_block, poll_block)
-
-        if reason ==17 or reason == 18: # Wimp Message
-            message = poll_block[4]
-            if message == 0:
-                _quit = True
+        reason,sender = swi.swi(
+            'Wimp_Poll','II;I.I',
+            0b1, ctypes.addressof(poll_buffer))
+
+        try:
+            poll_block = bytes(poll_buffer)
+            if reason == Wimp.ToolboxEvent:
+                size, reference, event_code, flags = \
+                    struct.unpack("IIII", poll_block[0:16])
+
+                if event_code == Toolbox.ObjectAutoCreated:
+                    name =''.join([chr(c) for c in \
+                        iter(lambda i=iter(poll_block[0x10:]): next(i), 0)])
+
+                    obj_class = swi.swi('Toolbox_GetObjectClass', '0I;I',
+                                        _id_block.self.id)
+
+                    print("auto-create {} {}".format(name, obj_class))
```

Comment:
Debug code left in? I'd usually use the logging interfaces or `if self.debug: ...` to make sure it isn't output normally.